### PR TITLE
Enable installed extensions by default

### LINKED ISSUE #1553: Enable installed extension automatically

### PR DESCRIPTION
**Pull request by panuhorsmalahti**
Fixes #1553

### CODE REVIEW

File: src/renderer/components/+extensions/extensions.tsx [hunk 1]
```diff
@@ -91,11 +91,23 @@ export class Extensions extends React.Component {
...
+
           Notifications.ok(
             <p>Extension <b>{displayName}</b> successfully installed!</p>
           );
           this.extensionState.delete(id);
           this.installPath = "";
+
+          // Enable installed extensions by default.
+          // Timeout is required since we're modifying the observable inside reaction.
+          setTimeout(() => {
```
Comment (ixrock) [inline src/renderer/components/+extensions/extensions.tsx]: > // Timeout is required since we're modifying the observable inside reaction.

Sounds weird.. Observables could/should be modified from anywhere.. 
Maybe you need `runInAction` or mark a method with `@action` when updating something?

Comment (panuhorsmalahti) [reply src/renderer/components/+extensions/extensions.tsx]: It's definitely weird, and it's a bit difficult to track what's going on with all the various reactions and computed extensions and store interaction.

I tried `runInAction` and  `@action` but those didn't work.

Comment (panuhorsmalahti) [reply src/renderer/components/+extensions/extensions.tsx]: Digging deeper into this one, it seems that the extensions-store will disable it if done sync in the reaction. I'm trying to find a better solution.

Comment (panuhorsmalahti) [reply src/renderer/components/+extensions/extensions.tsx]: I removed the timeout but also had to remove the reaction to update state from store updates in extensions-store. I didn't find a reason for that reaction, and enabled state seems to work fine with it removed.

File: src/extensions/extensions-store.ts [hunk 2]
```diff
@@ -43,16 +43,6 @@ export class ExtensionsStore extends BaseStore<LensExtensionsStoreModel> {
     await extensionLoader.whenLoaded;
     await this.whenLoaded;
 
-    // apply state on changes from store
```
Comment (nevalla) [inline src/extensions/extensions-store.ts:48]: Why are these lines removed? Removing this reaction seems to break enabling/disabling extensions from UI

Comment (panuhorsmalahti) [reply src/extensions/extensions-store.ts:48]: Removing these lines makes it possible to set isEnabled = true in the reaction. I didn't see any bugs from this change at least in my testing..

It would be nice for someone to explain the logic of why we need these lines.

Comment (panuhorsmalahti) [reply src/extensions/extensions-store.ts:48]: Okay, this does break the disabling to an extent. I'll change the PR yet again.

File: src/renderer/components/+extensions/extensions.tsx (hunk 1 above)
Comment (panuhorsmalahti) [reply src/renderer/components/+extensions/extensions.tsx]: This should be fixed now.

File: src/extensions/extensions-store.ts (hunk 2 above)
Comment (panuhorsmalahti) [reply src/extensions/extensions-store.ts:48]: PR fixed, enable and disable should work now.

File: src/extensions/extension-loader.ts [hunk 3]
```diff
@@ -25,7 +26,12 @@ const logModule = "[EXTENSIONS-LOADER]";
 export class ExtensionLoader {
   protected extensions = observable.map<LensExtensionId, InstalledExtension>();
   protected instances = observable.map<LensExtensionId, LensExtension>();
-  protected readonly requestExtensionsChannel = "extensions:loaded";
+
+  // IPC channel to broadcast changes to extensions from main
+  protected readonly extensionsMainChannel = "extensions:main";
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
  protected static readonly extensionsMainChannel = "extensions:main";
```

File: src/extensions/extension-loader.ts [hunk 4]
```diff
@@ -25,7 +26,12 @@ const logModule = "[EXTENSIONS-LOADER]";
...
+
+  // IPC channel to broadcast changes to extensions from renderer
+  protected readonly extensionsRendererChannel = "extensions:renderer";
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
  protected static readonly extensionsRendererChannel = "extensions:renderer";
```

File: src/extensions/extension-loader.ts [hunk 5]
```diff
@@ -95,43 +101,53 @@ export class ExtensionLoader {
     this.loadOnMain();
     this.broadcastExtensions();
 
-    reaction(() => this.extensions.toJS(), () => {
+    reaction(() => this.toJSON(), () => {
       this.broadcastExtensions();
     });
 
-    handleRequest(this.requestExtensionsChannel, () => {
+    handleRequest(this.extensionsMainChannel, () => {
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
    handleRequest(ExtensionLoader.extensionsMainChannel, () => {
```

File: src/extensions/extension-loader.ts [hunk 6]
```diff
@@ -95,43 +101,53 @@ export class ExtensionLoader {
...
       return Array.from(this.toJSON());
     });
+
+    subscribeToBroadcast(this.extensionsRendererChannel, (_event, extensions: [LensExtensionId, InstalledExtension][]) => {
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
    subscribeToBroadcast(ExtensionLoader.extensionsRendererChannel, (_event, extensions: [LensExtensionId, InstalledExtension][]) => {
```

File: src/extensions/extension-loader.ts [hunk 7]
```diff
@@ -95,43 +101,53 @@ export class ExtensionLoader {
...
       });
     };
 
-    requestMain(this.requestExtensionsChannel).then(extensionListHandler);
-    subscribeToBroadcast(this.requestExtensionsChannel, (event, extensions: [LensExtensionId, InstalledExtension][]) => {
+    reaction(() => this.toJSON(), () => {
+      this.broadcastExtensions(false);
+    });
+
+    requestMain(this.extensionsMainChannel).then(extensionListHandler);
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
    requestMain(ExtensionLoader.extensionsMainChannel).then(extensionListHandler);
```

File: src/extensions/extension-loader.ts [hunk 8]
```diff
@@ -95,43 +101,53 @@ export class ExtensionLoader {
...
+    subscribeToBroadcast(this.extensionsMainChannel, (_event, extensions: [LensExtensionId, InstalledExtension][]) => {
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
    subscribeToBroadcast(ExtensionLoader.extensionsMainChannel, (_event, extensions: [LensExtensionId, InstalledExtension][]) => {
```

File: src/extensions/extension-loader.ts [hunk 9]
```diff
@@ -151,7 +167,6 @@ export class ExtensionLoader {
   }
 
   loadOnClusterManagerRenderer() {
-    logger.info(`${logModule}: load on main renderer (cluster manager)`);
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts:154]: Please don't remove these, I think that it is valuable logging information. (If you must, then change it to debug)

File: src/extensions/extension-loader.ts [hunk 10]
```diff
@@ -257,8 +271,8 @@ export class ExtensionLoader {
     });
   }
 
-  broadcastExtensions() {
-    broadcastMessage(this.requestExtensionsChannel, Array.from(this.toJSON()));
+  broadcastExtensions(main = true) {
+    broadcastMessage(main ? this.extensionsMainChannel : this.extensionsRendererChannel, Array.from(this.toJSON()));
```
Comment (Nokel81) [inline src/extensions/extension-loader.ts]:
```suggestion
    broadcastMessage(main ? ExtensionLoader.extensionsMainChannel : ExtensionLoader.extensionsRendererChannel, Array.from(this.toJSON()));
```

File: src/extensions/extensions-store.ts [hunk 11]
```diff
@@ -61,7 +51,10 @@ export class ExtensionsStore extends BaseStore<LensExtensionsStoreModel> {
 
   isEnabled(extId: LensExtensionId) {
     const state = this.state.get(extId);
-    return state && state.enabled; // by default false
+
+    // By default false, so that copied extensions are disabled by default.
+    // If user installs the extension from the UI, the Extensions component will specifically enable it.
+    return Boolean(state && state.enabled);
```
Comment (Nokel81) [inline src/extensions/extensions-store.ts]:
```suggestion
    return !!state?.enabled;
```

**Review by Nokel81 — APPROVED** [review]

File: src/extensions/extensions-store.ts (hunk 11 above)
Comment (panuhorsmalahti) [reply src/extensions/extensions-store.ts]: I think `Boolean` is more readable than using `!!`. I think `?.` / `&&` are equivalent in readability.

File: src/extensions/extension-loader.ts (hunk 9 above)
Comment (panuhorsmalahti) [reply src/extensions/extension-loader.ts:154]: Added them back as .debug

File: src/extensions/extension-loader.ts (hunk 3 above)
Comment (panuhorsmalahti) [reply src/extensions/extension-loader.ts]: Changed

File: src/extensions/extension-loader.ts (hunk 4 above)
Comment (panuhorsmalahti) [reply src/extensions/extension-loader.ts]: Changed

File: src/extensions/extensions-store.ts (hunk 11 above)
Comment (Nokel81) [reply src/extensions/extensions-store.ts]: I a sort of moot on `Boolean` since it doesn't seem to be used much. I quite prefer `?.` over `&&` as the later can get very long quite quickly.

Comment (panuhorsmalahti) [reply src/extensions/extensions-store.ts]: I converted to `?.` but kept `Boolean` for now. It's true some don't know it, but I think it's pretty standard JavaScript, and every project is an opportunity to make people familiar with it. Quick look at Google seems to favor `Boolean` over `!!`.

**Review by jakolehm — APPROVED** [review]
LGTM
